### ISSUE TITLE
Having a preinstall script should not cause npm to skip node-gyp rebuild

### ISSUE THREAD

**Issue by gabrielschulhof** [npm/npm#12133]
In #2629, @mscdex found that adding a preinstall script to the package.json causes npm to also skip installation via `node-gyp rebuild`. I don't believe that his workaround of manually adding `node-gyp rebuild` as the install script is advisable, because that causes maintenance headaches down the line. What if, later on, npm decides that the default install script shall be something other than `node-gyp rebuild`?

The correct solution is to execute the custom preinstall script if one is found followed by the default install script if no custom install script is found.

Thus, in lieu of reopening #2629 I am opening a new issue.

**Comment by othiym23** [comment]
> The correct solution is to execute the custom preinstall script if one is found followed by the default install script if no custom install script is found.

This describes the current behavior. Do you mean that the default preinstall script should be run _in addition to_ the package's preinstall script?

**Comment by gabrielschulhof** [comment]
@othiym23 I do not mean that the default ```preinstall``` script should be run in addition to the custom ```preinstall``` script. What I'm observing even with the latest version of ```npm``` (3.8.6) is that the presence of a custom ```preinstall``` script causes the default ```install``` script to not run at all. To illustrate, consider the following package:

package.json:
```JSON
{
  "name": "hello",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "//": "The line below is problematic",
    "preinstall": "echo hello",
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "author": "",
  "license": "ISC",
  "dependencies": {
    "bindings": "^1.2.1"
  }
}
```
binding.gyp:
```JSON
{
	"targets": [ {
		"target_name": "hello",
		"sources": [ "hello.cc" ]
	} ]
}
```
hello.cc:
```C++
#include <node.h>
#include <v8.h>

using namespace v8;

void Init(Handle<Object> exports, Handle<Object> module) {}

NODE_MODULE(hello, Init)
```
and the following step:
```sh
npm install && node -p 'require( "bindings" )( "hello" );'
```
At this point, the **expected** behaviour is that ```node``` prints out the binary ```hello.node``` as an object and quits.

Unfortunately, the **actual** behaviour is that ```node``` cannot find the binary ```hello.node```, so it prints an exception and quits.

However, if you remove the line
```JSON
    "preinstall": "echo hello",
```
from ```package.json``` above (located just below the comment) and you repeat the step, you will have the expected behaviour.

In conclusion, having a custom **preinstall** script specified in the ```package.json``` causes ```npm``` to skip the *default* **install** script.

**Comment by gabrielschulhof** [comment]
I've tracked it down a bit and it boils down to an IMO incorrect assumption in normalize-package-data:
https://github.com/npm/normalize-package-data/blame/master/lib/normalize.js#L25-L29

**Comment by gabrielschulhof** [comment]
@myrne might be able to help us out.

**Comment by myrne** [comment]
@gabrielschulhof I'm sorry. I can't meaningfully comment on this. I don't recall writing the particular lines you mentioned. I assumed I must have copied them verbatim from read-package-json (from which normalize-package-data was extracted). I can't track down the original code in read-package-json though. Maybe it rings a bell with @isaacs .

**Comment by gabrielschulhof** [comment]
@othiym23 I believe I have isolated the places where the code makes the assumption that, if a package has a preinstall script then it is not a package with a gypfile. I have filed the above two bugs against each package wherein I have seen this assumption being made. I have locally tested the effect of removing the assumption and, indeed, it fixes the build process for my package. Unfortunately, I do not know the correct procedure for submitting a PR against multiple packages at the same time, while ensuring that Travis will correctly test the changes made to each of the packages.
